### DISCUSSION TITLE
Vercel is not supporting environment variables

### DISCUSSION THREAD

**Discussion by pushkarsingh019** [vercel/community#1338] [Help] [answered]
There have been countless discussions raised on this topic before, but somehow none of the earlier discussions have solved the problem.

## The Problem

Vercel is not recognising the environment variables.

I have added the environment variables using the vercel dashboard by going under the settings/environment variables.
<img width="933" alt="Screenshot 2023-01-15 at 6 25 04 PM" src="https://user-images.githubusercontent.com/94926273/212541902-90a527dd-728b-43d7-bae6-d718c1b2af3f.png">

I have re-deployed the project countless times, and I am accessing the environment variables in the following way

```
export const socketInUse = process.env.SOCKET || developmentSocket;
export const originInUse = process.env.ORIGIN || developmentOrigin;
```

somehow the error seems to exist after countless discussions. Just for reference, I am using create-react-app with the react of version `18.2`

Whenever I deploy the following code, it always goes to the developmentOrigin and developmentSocket and whenever I try to print the env variables it shows undefined. I tried adding the env variables in the same way, it works on other platforms such as Railway.app 

## Some solutions in earlier discussions

The main hack in earlier discussions was adding the env variables using the vercel CLI. This did work for some people, and some people are still facing the same issue.

Its always funny to notice that vercel is at the frontier of so many innovation in web, but this error someone seems to persist.

**Comment by socramm9** [comment, marked as answer]
I think you have to prefix  your env variable with `REACT_APP_` to work with create-react-app

In vercel and in code. So in your case it would be:

Vercel:
REACT_APP_ORIGIN
REACT_APP_SOCKET

and in code 
```
export const socketInUse = process.env.REACT_APP_SOCKET || developmentSocket;
export const originInUse = process.env.REACT_APP_ORIGIN || developmentOrigin;
```

**Reply by pushkarsingh019** [reply]
Thanks for the help.

Is this trick mentioned somewhere in the documentation that I might have missed? If not, how did you come across this hack?

**Reply by socramm9** [reply]
It is a create-react-app convention https://create-react-app.dev/docs/adding-custom-environment-variables/

Some other frameworks have other conversations. If you are using NextJS for example you would have to prefix your env with `NEXT_PUBLIC_`

**Reply by zohaibamir203** [reply]
I did same but still not working in my case

**Comment by gyl89** [comment]
If I am using expo sdk >=48, I need to prefix it with EXPO_PUBLIC (cf https://docs.expo.dev/guides/environment-variables/) but I also need to prefix with REACT_APP. Thus, my vercel environment variables are not seen by my react native web app

**Reply by jairaume** [reply]
Hi ! Did you find any solution for this, I have the same issue you have with a react native web app I'm trying to deploy on vercel.

**Comment by mostuselessboy** [comment]
instead of using the vercel user interface, just put all your secrets in .env.production, vercel will automatically read this and set it up.
![image](https://github.com/vercel/community/assets/85026406/d434a14a-8938-4154-af55-48ed33e20e6a)
